### PR TITLE
Placeholder and replace_component

### PR DESCRIPTION
**Pull request by jsn1993**
Stuff in this PR:
1. A major refactoring over the existing dsl code base to centralize all API implementations in Component. 
2. Implement Placeholder in dsl and add support in elaboration to recognize Placeholders and throw error in passes if there are leftover placeholders.
3. I started off implementing the delete_placeholder API but I realize that the delete_component is not that different lol. At the end I refactored add and delete into private functions and just expose replace_component to outside world.

### CODE REVIEW

**Comment by jsn1993** [comment]
I finished handling method ports and interfaces. Replacing a placeholder with a queue and let the queue take a parameter to configure himself to bypass/pipe looks very neat.

File: pymtl3/dsl/Component.py [hunk 1]
```diff
@@ -55,10 +61,542 @@ def _construct( s ):
...
+
+    for c in added_components:
+      top._collect_vars( c )
+
+    for c in added_signals | added_method_ports:
+      c._dsl.elaborate_top = top
+
+    # Lazy -- to avoid resolve_connection call which takes non-trivial
+    # time upon adding any connect, I just mark it here. Please make sure
+    # to call s.get_all_value_nets() to flush all pending connections
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: I think the pending connections will be resolved once you need it right? The comment here sounds like your have to manually call some API to resolve connections before using them...

File: pymtl3/dsl/Component.py [hunk 2]
```diff
@@ -55,10 +61,543 @@ def _construct( s ):
...
+
+    return _delete_component_internal( top, obj )
+    # import gc
+    # gc.collect() # this takes 0.1 seconds
+
+  #-----------------------------------------------------------------------
+  # Post-elaborate public APIs (can only be called after elaboration)
+  #-----------------------------------------------------------------------
+
+  """ Convenience/utility APIs """
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: I think docstring generally either appears right after a function/class definition or at the beginning of a module? Plus this doctoring will not be collected into `__doc__` of any methods...

File: pymtl3/dsl/Component.py [hunk 3]
```diff
@@ -55,10 +61,543 @@ def _construct( s ):
...
+        if is_list:
+          s._dsl.swapped_values[ component ] = ( current_obj, i, current_obj[i], is_list )
+          current_obj[i] = obj
+        else:
+          s._dsl.swapped_values[ component ] = ( current_obj, i, getattr(current_obj, i), is_list )
+          setattr( current_obj, i, obj )
+
+    s._dsl.locked_simulation = False
+
+  """ APIs that provide local metadata of a component """
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: Same here. I feel our `#-----` style separator works well in these kinds of situations. What do you think?

File: pymtl3/dsl/Component.py [hunk 4]
```diff
@@ -55,10 +61,543 @@ def _construct( s ):
...
+        iterable = enumerate( current_obj )
+        is_list = True
+      elif isinstance( current_obj, NamedObject ):
+        iterable = current_obj.__dict__.iteritems()
+        is_list = False
+      else:
+        return
+
+      for i, obj in iterable:
+        if not is_list and i.startswith("_"): # impossible to have tuple
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: `is_list` is constant within this for loop so you probably want to move the its check outside `for` to have better performance.

File: pymtl3/dsl/Component.py [hunk 5]
```diff
@@ -55,10 +61,543 @@ def _construct( s ):
...
+      parent._dsl.connect_order = new_connect_order
+
+      return saved_connections
+
+    return _delete_component_internal( top, obj )
+    # import gc
+    # gc.collect() # this takes 0.1 seconds
+
+  #-----------------------------------------------------------------------
+  # Post-elaborate public APIs (can only be called after elaboration)
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: I like the new API organization so much -- you don't have to dig into different levels to find out them!

File: pymtl3/dsl/Component.py [hunk 6]
```diff
@@ -55,10 +61,543 @@ def _construct( s ):
...
+    # call garbage collector. If we can make sure it is collected
+    # automatically and fast enough, we might remove this extra layer
+    #
+    # EDIT: After experimented w/ and w/o gc.collect(), it seems like it
+    # is eventually collected, but sometimes the intermediate memory
+    # footprint might reach up to gigabytes, so let's keep the
+    # gc.collect() for now
+
+    def _delete_component_internal( top, foo ):
+      is_placeholder = isinstance( obj, Placeholder )
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: Seems like you are not assuming `Placeholder` here.  What's this flag for?

File: pymtl3/dsl/ComponentLevel2.py [hunk 7]
```diff
@@ -451,7 +464,7 @@ def add_constraints( s, *args ): # add RD-U/WR-U constraints
           s._dsl.WR_U_constraints[ x0.var ].add( (sign, x1.func) )
 
   #-----------------------------------------------------------------------
-  # elaborate
+  # elaborate-related public APIs
```
Comment (ptpan) [inline pymtl3/dsl/ComponentLevel2.py]: Did you mean just `API` instead of `public API`?

File: pymtl3/dsl/Component.py (hunk 1 above)
Comment (jsn1993) [reply pymtl3/dsl/Component.py]: Good point. The comment is copied from the old implementation. What I actually mean is that whenever you call get_all_value_nets which is the API that gives you all the nets, the API itself will check if there are pending connections and flush them to return the latest nets.

File: pymtl3/dsl/Component.py (hunk 2 above)
Comment (jsn1993) [reply pymtl3/dsl/Component.py]: hmmm then what is a good seperator of blocks of APIs?

Comment (ptpan) [reply pymtl3/dsl/Component.py]: What do you think of `#--` here?

File: pymtl3/dsl/Component.py (hunk 6 above)
Comment (jsn1993) [reply pymtl3/dsl/Component.py]: Hmmm good catch. There is one line below that checks isinstance(obj,Placeholder) again. The purpose was to check if it is placeholder. If it is, I don't need to call those uncollect thingy.

File: pymtl3/dsl/ComponentLevel2.py (hunk 7 above)
Comment (jsn1993) [reply pymtl3/dsl/ComponentLevel2.py]: I guess I copy-paste a lot and now I need to fix them ...

File: pymtl3/dsl/Component.py (hunk 2 above)
Comment (jsn1993) [reply pymtl3/dsl/Component.py]: I want to make those "subsections" of the "Post-elaborate public APIs" section. In line 321-323 I already use #--. Any idea on a different one?

Comment (ptpan) [reply pymtl3/dsl/Component.py]: How about a single line `# Convenience/utility APIs` with a blank line before/after it? Or `## Convenience/utility APIs` like in markdown?

File: pymtl3/dsl/test/Placeholder_test.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    def construct( s ):
+      s.q = deque()
+      s.add_constraints( M(s.enq) < M(s.deq) )
+
+  a = FooCL_top()
+  a.elaborate()
+
+  a.replace_component( a.x.foo, RealQueue )
+
+  simple_sim_pass( a )
```
Comment (ptpan) [inline pymtl3/dsl/test/Placeholder_test.py]: Is it possible to have some preliminary assertions here? Just to make sure this works as intended.

Comment (jsn1993) [reply pymtl3/dsl/test/Placeholder_test.py]: hmm will do

File: pymtl3/dsl/Component.py (hunk 2 above)
Comment (jsn1993) [reply pymtl3/dsl/Component.py]: Well if it's single line, we have too many single line comments for each function in that part  ...

Comment (jsn1993) [reply pymtl3/dsl/Component.py]: @ptpan I guess we should spend sometime later to clean up all comment formats in one shot ...

**Pull request merged**